gnu/man/split.man
NAME
     split - split a file into pieces

SYNOPSIS
     split [-lines] [-l lines] [-b bytes[bkm]] [-C bytes[bkm]] [--lines=lines]
     [--bytes=bytes[bkm]] [--line-bytes=bytes[bkm]] [infile [outfile-prefix]]

DESCRIPTION
     This manual page documents the GNU version of split.  split  creates  one
     or  more  output  files  (as  many  as  necessary) containing consecutive
     sections of the infile, or the standard input if none  is  given  or  the
     name  `-' is given.  By default, split puts 1000 lines of the input file,
     or whatever is left if it is less than that, into each output file.

     The output file names consist of a prefix followed by a group of letters,
     chosen  so  that  concatenating  the output files in sorted order by file
     name produces the original input file, in order.  The default output file
     name  prefix is `x'.  If the outfile-prefix argument is given, it is used
     as the output file name prefix instead.

     OPTIONS

     -lines, -l lines, --lines=lines
          Put lines lines of the input file into each output file.

     -b bytes[bkm], --bytes=bytes[bkm]
          Put bytes bytes of the input file into each output file.  bytes is a
          nonzero  integer,  optionally  followed  by  one  of  the  following
          characters to specify a different unit.

     b    512-byte blocks.

     k    1-kilobyte blocks.

     m    1-megabyte blocks.

     -C bytes[bkm], --line-bytes=bytes[bkm]
          Put into each output file as many complete lines of the  input  file
          as  is  possible  without  exceeding bytes bytes.  If a line that is
          longer than bytes bytes occurs, put bytes  bytes  of  it  into  each
          output  file  until less than bytes bytes of the line are left, then
          continue normally.  bytes has the same format  as  for  the  --bytes
          option.

     The long-named options can be introduced with `+' as well  as  `--',  for
     compatibility with previous releases.  Eventually support for `+' will be
     removed, because it is incompatible with the POSIX.2 standard.


gnu/ChangeLog
Sun Feb  2 21:22:01 1992  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* Version 1.2.

	* nl.c: Support multiple files and "-" for stdin.
	(main): Check for read and write errors.
	(nl_file): New function.

Wed Jan 29 10:09:10 1992  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* tr.c (main): -t option was called -b in getopt spec.
	(validate): Don't warn that set1 is longer than set2.

	* tr.c: Rename --sysv-string2-truncate to --truncate-string1.

Fri Jan 17 16:29:05 1992  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* nl.c: New program from bin-src.

	* nl.c (main): Use a struct linebuffer for line_buf.
	(process_file): Use readline instead of fgets, to preserve NULs.
	(check_section): Use memcmp instead of strncmp.
	(proc_text): Print line_buf with fwrite instead of printf.

	* nl.c (main): Usage message if too many args given.  Check
	for error in closing input file.  Lengths of section delimiter
	strings were 1 too large.  Take separator_str into account in
	length of print_no_line_fmt.
	(build_print_fmt): Allocate space for print_fmt, in case
	separator_str is long.
	(proc_text): A blank line is one that contains nothing, not
	even nonprinting characters.

Fri Jan 17 01:04:22 1992  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* All programs: Document `--' instead of `+' to introduce
	long-named options, in usage messages.

	* sum.c (bsd_sum_file): Renamed from sum_file.
	(sysv_sum_file): New function.
	(main): Recognize an option to select between the algorithms.

Sun Jan  5 17:41:18 1992  Jim Meyering (meyering at churchy.gnu.ai.mit.edu)

	* pr.c (close_file, print_page): Fixed bug that had the command
	yes |head |pr -t printing "yyyyyyyyyy".
	* (print_page): Fixed bug that had pr -3 -a printing two too few
	trailer lines per page.
	* (main): Added restriction that -a and -m are incompatible.
	Although the POSIX spec doesn't explicitly say they shouldn't
	be used together, it says -a modifies the -column option and
	that -column shouldn't be used with -m.

Thu Jan  2 15:23:59 1992  David J. MacKenzie  (djm at albert.gnu.ai.mit.edu)

	* nl.c: Include regex.h after, not before, sys/types.h.

Thu Jan  2 12:18:10 1992  Tom Lord  (lord at geech.gnu.ai.mit.edu)

	* sort.c (fillbuf) return bytes buffered instead of bytes read.

Fri Dec 27 22:53:36 1991  Jim Kingdon  (kingdon at geech.gnu.ai.mit.edu)

	* sort.c (LINEALLOC): New #define.
	(struct lines): New field ``limit''.
	(initlines): Set it from new arg ``limit''.
	(sort, mergefps, checkfp): Pass new arg to initlines().
	(findlines): Don't realloc past lines->limit.

Tue Dec 24 01:24:03 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* tac.c, sort.c, csplit.c, system.h: Change POSIX ifdefs to
	HAVE_UNISTD_H and _POSIX_VERSION.

	* xwrite.c: Change POSIX ifdef to HAVE_UNISTD_H.

Sat 14 Dec 1991 11:46:42  Jim Meyering  (meyering at wombat)

	* tr.c: Fixed an inaccurate comment on posix_pedantic.

Thu 12 Dec 1991 21:15:20  Jim Meyering  (meyering at hal)

	* tr.c: Changed underscores to hyphens in long option name
	"sysv_string2_truncate".

Wed Dec 11 13:33:34 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* tac.c (main): Set obscure_syntax to tell re_search to
	allocate memory for the group registers.

Fri Dec  6 18:26:27 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* tac.c, sort.c, csplit.c [POSIX]:  Use sigaction instead of
	signal, which POSIX doesn't have.
	* sort.c: Replace inthandler and huphandler with sighandler.
	* csplit.c (main): Only handle signals if they were not being
	ignored.

	* tr.c: POSIX_ME_HARDER -> POSIXLY_CORRECT.

Wed Dec  4 00:47:47 1991  Jim Meyering  (meyering at wombat)

	* tr.c (unquote): Reformat code so it doesn't go beyond column 80.
	* tr.c (squeeze_filter): Comment a little on why it's better
	to step through the input by two.
	* tr.c (set_initialize): Write a comment describing the function.
	* tr.c: Eliminated the variable `portability_warnings' and replaced
	references to it by references to `!posix_pedantic'.  One of the
	uses of portability_warnings had been wrong.

Tue Dec  3 14:03:35 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* tr.c: New program.

Sun Dec  1 15:07:35 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* linebuffer.[ch] (freebuffer): New function (used by cron).

Thu Oct 17 22:30:22 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* system.h, configure, Makefile.in: Don't include memory.h if
	STDC_HEADERS, removing need for MEMORY_H_MISSING.

Thu 17 Oct 1991 16:53:07  Jim Meyering  (meyering at wombat)

	* pr.c (print_page): REALLY fixed `extra newline at EOF' problem.
	Somehow, part of my patch didn't make it last time.

Sat Oct 12 12:04:47 1991  David J. MacKenzie  (djm at churchy.gnu.ai.mit.edu)

	* tail.c (pipe_lines, pipe_bytes): Initialize `first->next'.

	* cmp.c (cmp): Print EOF message to stderr, not stdout, for
	POSIX 1003.2.11.2.

	* sort.c (xfwrite): fwrite never returns < 0, so check if
	number written is number we asked to write.
	(fillbuf, main): fread never returns < 0, so check ferror instead.
	From Rainer Orth.

Tue Oct  8 18:07:08 1991  Jim Meyering  (meyering at churchy)

	* pr.c (print_page): Really fixed `extra newline at EOF' problem.
	* (store_columns): Fixed bug that caused `pr -b -2' to coredump
	on files of certain lengths.

Fri Oct  4 22:30:25 1991  Jim Meyering  (meyering at churchy)

	* pr.c (print_page): Fixed to not add single spurious newline
	at EOF when using -t.

Wed Oct  2 01:02:05 1991  David J. MacKenzie  (djm at apple-gunkies)

	* pr.c (print_page): Don't pad the page if -t given.

	* csplit.c (load_buffer), sort.c (mergefps): Use bcopy, not memcpy.

	* memchr.c (memchr): Fix typo.

Thu Sep 26 12:35:00 1991  David J. MacKenzie  (djm at churchy.gnu.ai.mit.edu)


gnu/src/split.c
/* split.c -- split a file into pieces.
   Copyright (C) 1988, 1991 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.  */

/* By tege@sics.se, with rms.

   To do:
   * Implement -t CHAR or -t REGEX to specify break characters other
     than newline. */

#include <stdio.h>
#include <getopt.h>
#include <ctype.h>
#include <sys/types.h>
#include "system.h"

char *xmalloc ();
void error ();

int convint ();
int isdigits ();
int stdread ();
void line_bytes_split ();
void bytes_split ();
void cwrite ();
void lines_split ();
void next_file_name ();

/* Name under which this program was invoked.  */
char *program_name;

/* Base name of output files.  */
char *outfile;

/* Pointer to the end of the prefix in OUTFILE.
   Suffixes are inserted here.  */
char *outfile_mid;

/* Pointer to the end of OUTFILE. */
char *outfile_end;

/* Status for outfile name generation.  */
unsigned outfile_count = -1;
unsigned outfile_name_limit = 25 * 26;
unsigned outfile_name_generation = 1;

/* Name of input file.  May be "-".  */
char *infile;

/* Descriptor on which input file is open.  */
int input_desc;

/* Descriptor on which output file is open.  */
int output_desc;

void
usage (reason)
    char *reason;
{
  if (reason != NULL)
    fprintf (stderr, "%s: %s\n", program_name, reason);
  print_version();
  fprintf (stderr, "\
Usage: %s [-lines] [-l lines] [-b bytes[bkm]] [-C bytes[bkm]]\n\
       [--lines=lines] [--bytes=bytes[bkm]] [--line-bytes=bytes[bkm]]\n\
       [infile [outfile-prefix]]\n",
	   program_name);
  exit (2);
}

struct option longopts[] =
{
  {"bytes", 1, NULL, 'b'},
  {"lines", 1, NULL, 'l'},
  {"line-bytes", 1, NULL, 'C'},
  {NULL, 0, NULL, 0}
};

void
main (argc, argv)
    int argc;
    char *argv[];
{
  struct stat stat_buf;
  int num;			/* numeric argument from command line */
  enum
    {
      type_undef, type_bytes, type_byteslines, type_lines, type_digits
    } split_type = type_undef;
  int in_blk_size;		/* optimal block size of input file device */
  char *buf;			/* file i/o buffer */
  int accum = 0;
  char *outbase;
  int c;
  int digits_optind = 0;

  program_name = argv[0];

  /* Parse command line options.  */

  infile = "-";
  outbase = "x";

  while (1)
    {
      /* This is the argv-index of the option we will read next.  */
      int this_optind = optind ? optind : 1;

      c = getopt_long (argc, argv, "0123456789b:l:C:", longopts, (int *) 0);
      if (c == EOF)
	break;

      switch (c)
	{
	case 'b':
	  if (split_type != type_undef)
	    usage ("cannot split in more than one way");
	  split_type = type_bytes;
	  if (convint (optarg, &accum) == -1)
	    usage ("invalid number of bytes");
	  break;

	case 'l':
	  if (split_type != type_undef)
	    usage ("cannot split in more than one way");
	  split_type = type_lines;
	  if (!isdigits (optarg))
	    usage ("invalid number of lines");
	  accum = atoi (optarg);
	  break;

	case 'C':
	  if (split_type != type_undef)
	    usage ("cannot split in more than one way");
	  split_type = type_byteslines;
	  if (convint (optarg, &accum) == -1)
	    usage ("invalid number of bytes");
	  break;

	case '0':
	case '1':
	case '2':
	case '3':
	case '4':
	case '5':
	case '6':
	case '7':
	case '8':
	case '9':
	  if (split_type != type_undef && split_type != type_digits)
	    usage ("cannot split in more than one way");
	  if (digits_optind != 0 && digits_optind != this_optind)
	    accum = 0;		/* More than one number given; ignore other. */
	  digits_optind = this_optind;
	  split_type = type_digits;
	  accum = accum * 10 + c - '0';
	  break;

	default:
	  usage ((char *)0);
	}
    }

  /* Handle default case.  */
  if (split_type == type_undef)
    {
      split_type = type_lines;
      accum = 1000;
    }

  if (accum < 1)
    usage ("invalid number");
  num = accum;

  /* Get out the filename arguments.  */

  if (optind < argc)
    infile = argv[optind++];

  if (optind < argc)
    outbase = argv[optind++];

  if (optind < argc)
    usage ("too many arguments");

  /* Open the input file.  */
  if (!strcmp (infile, "-"))
  {
    if ( isatty(0) )
      usage (NULL);
    input_desc = 0;
  }
  else
    {
#ifdef OS2
      input_desc = open (infile, O_RDONLY | O_BINARY);
#else
      input_desc = open (infile, O_RDONLY);
#endif
      if (input_desc < 0)
	error (1, errno, "%s", infile);
    }

  /* No output file is open now.  */
  output_desc = -1;

  /* Copy the output file prefix so we can add suffixes to it.
     26**29 is certainly enough output files!  */

  outfile = xmalloc (strlen (outbase) + 30);
  strcpy (outfile, outbase);
  outfile_mid = outfile + strlen (outfile);
  outfile_end = outfile_mid + 2;
  bzero (outfile_mid, 30);
  outfile_mid[0] = 'a';
  outfile_mid[1] = 'a' - 1;  /* first call to next_file_name makes it an 'a' */

  /* Get the optimal block size of input device and make a buffer.  */

  if (fstat (input_desc, &stat_buf) < 0)
    error (1, errno, "%s", infile);
  in_blk_size = ST_BLKSIZE (stat_buf);

  buf = xmalloc (in_blk_size + 1);

  switch (split_type)
    {
    case type_digits:
    case type_lines:
      lines_split (num, buf, in_blk_size);
      break;

    case type_bytes:
      bytes_split (num, buf, in_blk_size);
      break;

    case type_byteslines:
      line_bytes_split (num);
      break;
    }

  if (close (input_desc) < 0)
    error (1, errno, "%s", infile);
  if (output_desc >= 0 && close (output_desc) < 0)
    error (1, errno, "%s", outfile);

  exit (0);
}

/* Return nonzero if the string STR is composed entirely of decimal digits.  */

int
isdigits (str)
    char *str;
{
  do
    {
      if (!isdigit (*str))
	return 0;
      str++;
    }
  while (*str);
  return 1;
}

/* Put the value of the number in STR into *VAL.
   STR can specify a positive integer, optionally ending in `k'
   to mean kilo or `m' to mean mega.
   Return 0 if STR is valid, -1 if not. */

int
convint (str, val)
     char *str;
     int *val;
{
  int multiplier = 1;
  int arglen = strlen (str);

  if (arglen > 1)
    {
      switch (str[arglen - 1])
	{
	case 'b':
	  multiplier = 512;
	  str[arglen - 1] = '\0';
	  break;
	case 'k':
	  multiplier = 1024;
	  str[arglen - 1] = '\0';
	  break;
	case 'm':
	  multiplier = 1048576;
	  str[arglen - 1] = '\0';
	  break;
	}
    }
  if (!isdigits (str))
    return -1;
  *val = atoi (str) * multiplier;
  return 0;
}

/* Split into pieces of exactly NCHARS bytes.
   Use buffer BUF, whose size is BUFSIZE.  */

void
bytes_split (nchars, buf, bufsize)
    int nchars;
    char *buf;
    int bufsize;
{
  int n_read;
  int new_file_flag = 1;
  int to_read;
  int to_write = nchars;
  char *bp_out;

  do
    {
      n_read = stdread (buf, bufsize);
      if (n_read < 0)
        error (1, errno, "%s", infile);
      bp_out = buf;
      to_read = n_read;
      for (;;)
	{
	  if (to_read < to_write)
	    {
	      if (to_read)	/* do not write 0 bytes! */
		{
		  cwrite (new_file_flag, bp_out, to_read);
		  to_write -= to_read;
		  new_file_flag = 0;
		}
	      break;
	    }
	  else
	    {
	      cwrite (new_file_flag, bp_out, to_write);
	      bp_out += to_write;
	      to_read -= to_write;
	      new_file_flag = 1;
	      to_write = nchars;
	    }
	}
    }
  while (n_read == bufsize);
}

/* Split into pieces of exactly NLINES lines.
   Use buffer BUF, whose size is BUFSIZE.  */

void
lines_split (nlines, buf, bufsize)
    int nlines;
    char *buf;
    int bufsize;
{
  int n_read;
  char *bp, *bp_out, *eob;
  int new_file_flag = 1;
  int n = 0;

  do
    {
      n_read = stdread (buf, bufsize);
      if (n_read < 0)
	error (1, errno, "%s", infile);
      bp = bp_out = buf;
      eob = bp + n_read;
      *eob = '\n';
      for (;;)
	{
	  while (*bp++ != '\n')
	    ;			/* this semicolon takes most of the time */
	  if (bp > eob)
	    {
	      if (eob != bp_out) /* do not write 0 bytes! */
		{
		  cwrite (new_file_flag, bp_out, eob - bp_out);
		  new_file_flag = 0;
		}
	      break;
	    }
	  else
	    if (++n >= nlines)
	      {
		cwrite (new_file_flag, bp_out, bp - bp_out);
		bp_out = bp;
		new_file_flag = 1;
		n = 0;
	      }
	}
    }
  while (n_read == bufsize);
}

/* Split into pieces that are as large as possible while still not more
   than NCHARS bytes, and are split on line boundaries except
   where lines longer than NCHARS bytes occur. */

void
line_bytes_split (nchars)
    int nchars;
{
  int n_read;
  char *bp;
  int eof = 0;
  int n_buffered = 0;
  char *buf = (char *) xmalloc (nchars);

  do
    {
      /* Fill up the full buffer size from the input file.  */

      n_read = stdread (buf + n_buffered, nchars - n_buffered);
      if (n_read < 0)
	error (1, errno, "%s", infile);

      n_buffered += n_read;
      if (n_buffered != nchars)
	eof = 1;

      /* Find where to end this chunk.  */
      bp = buf + n_buffered;
      if (n_buffered == nchars)
	{
	  while (bp > buf && bp[-1] != '\n')
	    bp--;
	}

      /* If chunk has no newlines, use all the chunk.  */
      if (bp == buf)
	bp = buf + n_buffered;

      /* Output the chars as one output file.  */
      cwrite (1, buf, bp - buf);

      /* Discard the chars we just output; move rest of chunk
	 down to be the start of the next chunk.  */
      n_buffered -= bp - buf;
      if (n_buffered > 0)
	bcopy (bp, buf, n_buffered);
    }
  while (!eof);
  free (buf);
}

/* Write BYTES bytes at BP to an output file.
   If NEW_FILE_FLAG is nonzero, open the next output file.
   Otherwise add to the same output file already in use.  */

void
cwrite (new_file_flag, bp, bytes)
    int new_file_flag;
    char *bp;
    int bytes;
{
  if (new_file_flag)
    {
      if (output_desc >= 0 && close (output_desc) < 0)
	error (1, errno, "%s", outfile);

      next_file_name ();
#ifdef OS2
      output_desc = open (outfile, O_WRONLY | O_CREAT | O_TRUNC | O_BINARY, 0666);
#else
      output_desc = open (outfile, O_WRONLY | O_CREAT | O_TRUNC, 0666);
#endif
      if (output_desc < 0)
	error (1, errno, "%s", outfile);
    }
  if (write (output_desc, bp, bytes) < 0)
    error (1, errno, "%s", outfile);
}

/* Read NCHARS bytes from the input file into BUF.
   Return the number of bytes successfully read.
   If this is less than NCHARS, do not call `stdread' again.  */

int
stdread (buf, nchars)
    char *buf;
    int nchars;
{
  int n_read;
  int to_be_read = nchars;

  while (to_be_read)
    {
      n_read = read (input_desc, buf, to_be_read);
      if (n_read < 0)
	return -1;
      if (n_read == 0)
	break;
      to_be_read -= n_read;
      buf += n_read;
    }
  return nchars - to_be_read;
}

/* Compute the next sequential output file name suffix and store it
   into the string `outfile' at the position pointed to by `outfile_mid'.  */

void
next_file_name ()
{
  int x;
  char *ne;

  outfile_count++;
  if (outfile_count < outfile_name_limit)
    {
      for (ne = outfile_end - 1; ; ne--)
	{
	  x = *ne;
	  if (x != 'z')
	    break;
	  *ne = 'a';
	}
      *ne = x + 1;
      return;
    }

  outfile_count = 0;
  outfile_name_limit *= 26;
  outfile_name_generation++;
  *outfile_mid++ = 'z';
  for (x = 0; x <= outfile_name_generation; x++)
    outfile_mid[x] = 'a';
  outfile_end += 2;
}
